File: src/main/java/com/magina/resource/config/BeansConfiguration.java
package com.magina.resource.config;

import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.thymeleaf.ThymeleafProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.Ordered;
import org.springframework.util.MimeType;
import org.springframework.web.servlet.ViewResolver;
import org.thymeleaf.spring4.SpringTemplateEngine;
import org.thymeleaf.spring4.resourceresolver.SpringResourceResourceResolver;
import org.thymeleaf.spring4.view.ThymeleafViewResolver;
import org.thymeleaf.templateresolver.TemplateResolver;

import com.google.common.collect.Maps;

import nz.net.ultraq.thymeleaf.LayoutDialect;

@Configuration
public class BeansConfiguration {

    @Autowired
    private ThymeleafProperties properties;

    @Bean
    public SpringResourceResourceResolver thymeleafResourceResolver() {
        return new SpringResourceResourceResolver();
    }

    @Bean
    public TemplateResolver templateResolver() {
        TemplateResolver resolver = new TemplateResolver();
        resolver.setResourceResolver(thymeleafResourceResolver());
        resolver.setPrefix(this.properties.getPrefix());
        resolver.setSuffix(this.properties.getSuffix());
        resolver.setTemplateMode(this.properties.getMode());
        if (this.properties.getEncoding() != null) {
            resolver.setCharacterEncoding(this.properties.getEncoding().name());
        }
        resolver.setCacheable(this.properties.isCache());
        Integer order = this.properties.getTemplateResolverOrder();
        if (order != null) {
            resolver.setOrder(order);
        }
        return resolver;
    }

    @Bean
    public SpringTemplateEngine templateEngine() {
        SpringTemplateEngine templateEngine = new SpringTemplateEngine();
        templateEngine.setTemplateResolver(templateResolver());
        templateEngine.addDialect(new LayoutDialect());
        return templateEngine;
    }

    @Bean
    public ViewResolver viewResolver() {
        ThymeleafViewResolver resolver = new CustomThymeleafViewResolver();
        resolver.setTemplateEngine(templateEngine());
        resolver.setCharacterEncoding(this.properties.getEncoding().name());
        resolver.setContentType(appendCharset(this.properties.getContentType(), resolver.getCharacterEncoding()));
        resolver.setExcludedViewNames(this.properties.getExcludedViewNames());
        resolver.setViewNames(this.properties.getViewNames());
        resolver.setOrder(Ordered.LOWEST_PRECEDENCE - 5);
        resolver.setCache(this.properties.isCache());
        return resolver;
    }

    private String appendCharset(MimeType type, String charset) {
        if (type.getCharset() != null) {
            return type.toString();
        }
        Map<String, String> parameters = Maps.newLinkedHashMap();
        parameters.put("charset", charset);
        parameters.putAll(type.getParameters());
        return new MimeType(type, parameters).toString();
    }

}

File: src/main/kotlin/com/magina/resource/config/BeansConfiguration.kt
package com.magina.resource.config

import nz.net.ultraq.thymeleaf.LayoutDialect
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.thymeleaf.ThymeleafProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.Ordered
import org.springframework.util.MimeType
import org.springframework.web.servlet.ViewResolver
import org.thymeleaf.spring4.SpringTemplateEngine
import org.thymeleaf.spring4.resourceresolver.SpringResourceResourceResolver
import org.thymeleaf.templateresolver.TemplateResolver
import java.util.LinkedHashMap

@Configuration
open class BeansConfiguration {

	@Autowired
	private var properties: ThymeleafProperties? = null

	@Bean
	open fun thymeleafResourceResolver(): SpringResourceResourceResolver {
		return SpringResourceResourceResolver()
	}

	@Bean
	open fun templateResolver(): TemplateResolver {
		var resolver = TemplateResolver()
		resolver.setResourceResolver(thymeleafResourceResolver())
		if (properties != null) {
			resolver.setPrefix(properties?.getPrefix())
			resolver.setSuffix(properties?.getSuffix())
			resolver.setTemplateMode(properties?.getMode())
			if (properties?.getEncoding() != null) {
				resolver.setCharacterEncoding(properties?.encoding?.name())
			}
			var cache = properties?.isCache
			if (cache != null) {
				resolver.setCacheable(cache)
			}
			var order = properties?.templateResolverOrder
			if (order != null) {
				resolver.setOrder(order)
			}
		}
		return resolver
	}

	@Bean
	open fun templateEngine(): SpringTemplateEngine {
		var templateEngine = SpringTemplateEngine()
		templateEngine.setTemplateResolver(templateResolver())
		templateEngine.addDialect(LayoutDialect())
		return templateEngine
	}

	@Bean
	open fun viewResolver(): ViewResolver {
		var resolver = CustomThymeleafViewResolver()
		resolver.setTemplateEngine(templateEngine())
		resolver.setOrder(Ordered.LOWEST_PRECEDENCE - 5)
		if (properties != null) {
			resolver.setCharacterEncoding(properties?.getEncoding()?.name())
			resolver.setContentType(appendCharset(properties?.getContentType(), resolver.getCharacterEncoding()))
			resolver.setExcludedViewNames(properties?.getExcludedViewNames())
			resolver.setViewNames(properties?.getViewNames())
			var cache = properties?.isCache()
			if (cache != null) resolver.setCache(cache)
		}
		return resolver
	}

	private fun appendCharset(type: MimeType?, charset: String): String {
		if (type != null && type.getCharset() != null) {
			return type.toString()
		}
		var parameters = LinkedHashMap<String, String>()
		parameters.put("charset", charset)
		parameters.putAll(type?.getParameters() ?: LinkedHashMap<String, String>())
		return MimeType(type, parameters).toString()
	}

}
